File: Python/621_TaskScheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        # A dictionary with each task..
        # Get a task from the dictionary every time I 'can'
        # Otherwise wait.
        
        taskInformation = dict()
        for task in tasks:
            if task in taskInformation: 
                taskInformation[task][0] = taskInformation[task][0] + 1 
            else: taskInformation[task] = [1,0]
        
        remaining = len(taskInformation)
        clock = 0
        while remaining:
            modified = -1
            for task,info in taskInformation.items():
                if info[1] <= clock:
                    modified = task
                    break;
            if modified != -1:
                taskInformation[modified][0] -= 1
                taskInformation[modified][1] += (n+1)
                if taskInformation[modified][0] == 0: 
                    taskInformation.pop(modified)
                    remaining = len(taskInformation)
                
            clock += 1
        return clock
        

File: C#/621_TaskScheduler.cs
public class Solution {

    // The idea here is to use a list to optimize tasks that need to run. 
    // We save the last time it was executed and peek it until ready to run again. 
    // After running we send it to the back of the list.
    public int LeastInterval(char[] tasks, int n) {
        // An auxiliary dictionary..
        Dictionary<char,(int,int)> taskInfo = new Dictionary<char,(int,int)>();
        foreach (char c in tasks) {
            if (!taskInfo.TryAdd(c, (1,0)))
                taskInfo[c] = (taskInfo[c].Item1 + 1, taskInfo[c].Item2);
        }
        // We will use this to itarate faster!
        Queue<(int,int)> remaining = new Queue<(int, int)>();
        foreach (var kvp in taskInfo) { 
            remaining.Enqueue(kvp.Value);
        }
            
        int clock = 0;
        while (remaining.Count > 0) {
            char found = '\0';
            if (remaining.Peek().Item2 <= clock) {
                (int quantity, int nextClock) = remaining.Dequeue();
                if (quantity > 1){
                    remaining.Enqueue((quantity - 1, nextClock + n + 1));
                }
            }
            clock++;
        }
        return clock;
    }
}
